mongo/client_bulk_write.go

```diff
@@ -0,0 +1,718 @@
+		if exception == nil {
+			exception = new(ClientBulkWriteException)
+		}
+		exception.WriteConcernErrors = batches.writeConcernErrors
+		exception.WriteErrors = batches.writeErrors
+	}
+	if exception != nil {
+		var hasSuccess bool
+		if batches.ordered {
+			_, ok := batches.writeErrors[0]
+			hasSuccess = !ok
+		} else {
+			hasSuccess = len(batches.writeErrors) < len(bw.models)
+		}
+		if hasSuccess {
+			exception.PartialResult = batches.result
+		}
+		return *exception
+	}
+	return err
+}
+
+func (bw *clientBulkWrite) newCommand() func([]byte, description.SelectedServer) ([]byte, error) {
+	return func(dst []byte, desc description.SelectedServer) ([]byte, error) {
+		dst = bsoncore.AppendInt32Element(dst, "bulkWrite", 1)
+
+		dst = bsoncore.AppendBooleanElement(dst, "errorsOnly", bw.errorsOnly)
+		if bw.bypassDocumentValidation != nil && (desc.WireVersion != nil && driverutil.VersionRangeIncludes(*desc.WireVersion, 4)) {
+			dst = bsoncore.AppendBooleanElement(dst, "bypassDocumentValidation", *bw.bypassDocumentValidation)
+		}
+		if bw.comment != nil {
+			comment, err := marshalValue(bw.comment, bw.client.bsonOpts, bw.client.registry)
+			if err != nil {
+				return nil, err
+			}
+			dst = bsoncore.AppendValueElement(dst, "comment", comment)
+		}
+		dst = bsoncore.AppendBooleanElement(dst, "ordered", bw.ordered == nil || *bw.ordered)
+		if bw.let != nil {
+			let, err := marshal(bw.let, bw.client.bsonOpts, bw.client.registry)
+			if err != nil {
+				return nil, err
+			}
+			dst = bsoncore.AppendDocumentElement(dst, "let", let)
+		}
+		return dst, nil
+	}
+}
+
+type cursorInfo struct {
+	Ok        bool
+	Idx       int32
+	Code      *int32
+	Errmsg    *string
+	ErrInfo   bson.Raw
+	N         int32
+	NModified *int32
+	Upserted  *struct {
+		ID interface{} `bson:"_id"`
+	}
+}
+
+func (cur *cursorInfo) extractError() *WriteError {
+	if cur.Ok {
+		return nil
+	}
+	err := &WriteError{
+		Index:   int(cur.Idx),
+		Details: cur.ErrInfo,
+	}
+	if cur.Code != nil {
+		err.Code = int(*cur.Code)
+	}
+	if cur.Errmsg != nil {
+		err.Message = *cur.Errmsg
+	}
+	return err
+}
+
+type modelBatches struct {
+	session *session.Client
+	client  *Client
+
+	ordered bool
+	models  []clientWriteModel
+
+	offset int
+
+	retryMode      driver.RetryMode // RetryNone by default
+	cursorHandlers []func(*cursorInfo, bson.Raw) bool
+	newIDMap       map[int]interface{}
+
+	result             *ClientBulkWriteResult
+	writeConcernErrors []WriteConcernError
+	writeErrors        map[int]WriteError
+}
+
+func (mb *modelBatches) IsOrdered() *bool {
+	return &mb.ordered
+}
+
+func (mb *modelBatches) AdvanceBatches(n int) {
+	mb.offset += n
+	if mb.offset > len(mb.models) {
+		mb.offset = len(mb.models)
+	}
+}
+
+func (mb *modelBatches) Size() int {
+	if mb.offset > len(mb.models) {
+		return 0
+	}
+	return len(mb.models) - mb.offset
+}
+
+func (mb *modelBatches) AppendBatchSequence(dst []byte, maxCount, maxDocSize, totalSize int) (int, []byte, error) {
```

Comment:
Do we have coverage for this in the following two cases:

1. One doc exceeds `maxBsonObjectSize`, I would expect an error.
2. Multiple documents < `maxBsonObjectSize` together exceed `maxMessageSizeBytes`. I would expect n > 1.

Here is a helpful gist I used: https://gist.github.com/prestonvasquez/40daa306abc5a21147235d2ea982d841

Reply:
I've involved GODRIVER-3421 in fc91428, so the first case is no longer needed. The second case is added in both [client_bulk_write_test.go](https://github.com/mongodb/mongo-go-driver/pull/1884/commits/fc91428893d9021d90c8ea9f1128a60f9f5c976a#diff-a31473171826e5bdf71d4f0f78929323a760cd15d1a4588608a783f6bec30ab6R102) and [client_test.go](https://github.com/mongodb/mongo-go-driver/pull/1884/commits/fc91428893d9021d90c8ea9f1128a60f9f5c976a#diff-4220cfa18a0cd4b8c2753f5a72e2f1f6645593197b897033c3631809eb888d13R520).